DirectProgramming/DPC++FPGA/Tutorials/Features/floating_point_pragmas/src/fp_pragmas.cpp

```diff
@@ -0,0 +1,304 @@
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only);
+      accessor accessorArray(bufferArray, h, read_only);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ContractFastKernel>([=]() [[intel::kernel_args_restrict]] {
+        accessorRes[0] = 0.0;
+
+        for (size_t i = 0; i < size; i++) {
+          double temp1 = 0.0, temp2 = 0.0;
+          temp1 = accessorA[0] + accessorB[0];
+          temp2 = accessorC[0] + accessorArray[i];
+          accessorRes[0] += temp1 * temp2 + accessorD[0];
+        }
+      });
+    });
+  } catch (sycl::exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
+      std::cerr << "If you are targeting an FPGA, please ensure that your "
+                   "system has a correctly configured FPGA board.\n";
+      std::cerr << "Run sys_check in the oneAPI root directory to verify.\n";
+      std::cerr << "If you are targeting the FPGA emulator, compile with "
+                   "-DFPGA_EMULATOR.\n";
+    }
+    std::terminate();
+  }
+}
+
+// Launch a kernel on the device specified by selector.
+// The kernel's functionality is designed to show the
+// performance impact of the fp contract pragma.
+void disable_contract(const device_selector &selector, const double a,
+                      const double b, const double c, const double d,
+                      const double *arr, const size_t size,
+                      double &res) {
+  try {
+    queue q(selector, dpc_common::exception_handler,
+            property::queue::enable_profiling{});
+
+    buffer<double, 1> bufferA(&a, 1);
+    buffer<double, 1> bufferB(&b, 1);
+    buffer<double, 1> bufferC(&c, 1);
+    buffer<double, 1> bufferD(&d, 1);
+    buffer<double, 1> bufferE(&res, 1);
+    buffer<double, 1> bufferArray(arr, size);
+
+    event e = q.submit([&](handler &h) {
+      accessor accessorA(bufferA, h, read_only);
+      accessor accessorB(bufferB, h, read_only);
+      accessor accessorC(bufferC, h, read_only);
+      accessor accessorD(bufferD, h, read_only);
+      accessor accessorRes(bufferE, h, write_only);
+      accessor accessorArray(bufferArray, h, read_only);
+
+      // The "kernel_args_restrict" tells the compiler that a, b, and r
+      // do not alias. For a full explanation, see:
+      //    DPC++FPGA/Tutorials/Features/kernel_args_restrict
+      h.single_task<ContractOffKernel>([=]() [[intel::kernel_args_restrict]] {
+#pragma clang fp contract(off)
+
+        accessorRes[0] = 0.0;
+
+        for (size_t i = 0; i < size; i++) {
```

Comment:
Rewrited to refactor code using boolean template. However, kernel lambda could not be refactored because different pragmas need to apply to the lambda function.